Internal/svTrainNN.ecl

```diff
@@ -0,0 +1,494 @@
+              wUpdates_[windx] += wUpdate;
+            }
+          }
+        }
+      }
+      /**
+        * Train the neural network and output the weight updates.
+        * Runs all of the training samples.
+        */
+      void Train(tp * train, uint32_t trainSize)
+      {
+        t_ = train;
+        tsize_ = trainSize;
+        for (uint32_t i = 0; i < trainSize;i++)
+        {
+          // Do the Forward and Back Propagation to train the weights
+          fbPropagate(train[i]);
+        }
+      }
+      // NN Destructor
+      ~NeurNet()
+      {
+        rtlFree(hidden_);
+      }
+  };
+  /**
+    * Process the returned stramed dataset, one slice at a time.
+    * Before returning the first slice, retrieve the weights and the
+    * training data from the input streams, and then run the neural
+    * network to update the weights.  The returned dataset contains
+    * the weight updates (i.e. deltas), and not the final weights.
+    */
+  class MyStreamInlineDataset : public RtlCInterface, implements IRowStream
+  {
+    double * fW_; // Flattened weights
+    double * fU_; // Updated weights
+    tp* fT_;     // Flattened training data
+    uint32_t fT_count_;  // Number of training records
+    NeurNet *nn_;       // Neural network instance
+    bool calculated_ = false; // Flag so that we know we've already
+                              // calculated the weight updates (i.e. before
+                              // returning the first slice.
+    uint16_t nodeId_ = 0; // Current node #
+    uint16_t outSlice_ = 0; // Current outSlice index.
+    uint32_t weightCount_; // The total number of weights
+    float loss_; // The computed Loss
+    float minLoss_; // The lowest lost we've seen so far (passthrough)
+    uint32_t minEpoch_; // Epoch of minimum loss
+    uint32_t maxNoProg_; // Maximum epochs with no progress (passthrough)
+    uint64_t batchPos_; // The current position within the full training set (passthrough)
+  public:
+    MyStreamInlineDataset(IEngineRowAllocator * _resultAllocator, IRowStream * _wts, IRowStream * _train,
+          uint32_t _slicesize, uint32_t _nweights,
+          uint32_t _numwords, uint32_t _dim, uint32_t _mbsize, float _lr, uint32_t _negsamp)
+         : resultAllocator(_resultAllocator), wts(_wts), train(_train), slicesize(_slicesize),
+                           nweights(_nweights),
+                           numwords(_numwords), dim(_dim), mbsize(_mbsize), negsamp(_negsamp), lr(_lr)
+    {
+       weightCount_ = nweights;
+       // Allocate space for the weights and the weight updates
+       fW_ = (double *)rtlMalloc(weightCount_ * sizeof(double)); // Flattened weight array
+       fU_ = (double *)rtlMalloc(weightCount_ * sizeof(double)); // Flattened weight update array
+       // Initialize weight updates to zero
+       for (uint32_t i = 0; i < weightCount_;i++)
+          fU_[i] = 0.0;
+       fT_ = (tp *)rtlMalloc(mbsize * sizeof(tp)); // Flattened training data array
```

Comment:
need to call rtlCalloc (new function) or zero the array - otherwise if the destructor is called before nextRow() it will attempt to access uninitialised memory (fT_[x].contx).